src/Language/Haskell/Refact/Utils/ExactPrint.hs

```diff
@@ -160,3 +177,146 @@ balanceAllComments la
       return t
 
 -- ---------------------------------------------------------------------
+--Useful helper function that logs the current refact parsed
+
+exactPrintParsed :: RefactGhc ()
+exactPrintParsed = do
+  parsed <- getRefactParsed
+  anns <- fetchAnnsFinal
+  let str = exactPrint parsed anns
+  logm str
+
+-- ---------------------------------------------------------------------
+--A helper function that logs chunks of ast
+
+exactPrintExpr :: Annotate ast => GHC.Located ast -> RefactGhc ()
+exactPrintExpr ast = do
+  anns <- fetchAnnsFinal
+  let str = exactPrint ast anns
+  logm str
+
+-- ---------------------------------------------------------------------
+
+--This generates a unique location and wraps the given ast chunk with that location
+--Also adds an empty annotation at that location
+locate :: (SYB.Data a) => a -> RefactGhc (GHC.Located a)
+locate ast = do
+  loc <- liftT uniqueSrcSpanT
+  let res = (GHC.L loc ast)
+  addEmptyAnn res
+  return res
+
+--Adds an empty annotation at the provided location
+addEmptyAnn :: (SYB.Data a) => GHC.Located a -> RefactGhc ()
+addEmptyAnn a = addAnn a annNone
+
+--Adds an "AnnVal" annotation at the provided location
+addAnnVal :: (SYB.Data a) => GHC.Located a -> RefactGhc ()
+addAnnVal a = addAnn a valAnn
+  where valAnn = annNone {annEntryDelta = DP (0,1), annsDP = [(G GHC.AnnVal, DP (0,0))]}
+
+--Adds the given annotation at the provided location
+addAnn :: (SYB.Data a) => GHC.Located a -> Annotation -> RefactGhc ()
+addAnn a ann = do
+  currAnns <- fetchAnnsFinal
+  let k = mkAnnKey a
+  setRefactAnns $ Map.insert k ann currAnns
+
+--Sets the entry delta position of an ast chunk
+setDP :: (SYB.Data a) => DeltaPos -> GHC.Located a -> RefactGhc ()
+setDP dp ast = do
+  currAnns <- fetchAnnsFinal
+  let k = mkAnnKey ast
+      mv = Map.lookup k currAnns
+  case mv of
+    Nothing -> return ()
+    Just v -> addAnn ast (v {annEntryDelta = dp})
+
+--Resets the given AST chunk's delta position to zero.
+zeroDP :: (SYB.Data a) => GHC.Located a -> RefactGhc ()
+zeroDP = setDP (DP (0,0))
+
+--This just pulls out the successful result from an exact print parser or throws an error if the parse was unsuccessful.
+handleParseResult :: String -> Either (GHC.SrcSpan, String) (Anns, a) -> RefactGhc (Anns, a)
+handleParseResult msg e = case e of
+  (Left (_, errStr)) -> error $ "The parse from: " ++ msg ++ " with error: " ++ errStr
+  (Right res) -> return res
+
+-- Retrieves all annotations that correspond to all subtrees of the provided ast chunk
+getAllAnns :: (SYB.Data a) => Anns -> a -> Anns
```

Comment:
This looks like a debug display routine.

When I am developing, I make heavy use of `logDataWithAnns` which will dump an AST fragment with the annotations interleaved. FYI